server/src/main/java/com/thoughtworks/go/server/service/AgentService.java

```diff
@@ -126,143 +136,405 @@ public AgentsViewModel registeredAgents() {
-            agentConfigService.updateAgentIpByUuid(agentConfig.getUuid(), info.getIpAdress(), userName);
+    }
+
+    private void bombIfAgentDoesNotHaveACookie(AgentRuntimeInfo agentRuntimeInfo) {
+        if (!agentRuntimeInfo.hasCookie()) {
+            LOGGER.warn("Agent [{}] has no cookie set", agentRuntimeInfo.agentInfoDebugString());
+            throw new AgentNoCookieSetException(format("Agent [%s] has no cookie set", agentRuntimeInfo.agentInfoDebugString()));
         }
-        agentInstances.updateAgentRuntimeInfo(info);
     }
 
     public Username agentUsername(String uuId, String ipAddress, String hostNameForDisplay) {
-        return new Username(String.format("agent_%s_%s_%s", uuId, ipAddress, hostNameForDisplay));
+        return new Username(format("agent_%s_%s_%s", uuId, ipAddress, hostNameForDisplay));
     }
 
     public Registration requestRegistration(Username username, AgentRuntimeInfo agentRuntimeInfo) {
         LOGGER.debug("Agent is requesting registration {}", agentRuntimeInfo);
+
         AgentInstance agentInstance = agentInstances.register(agentRuntimeInfo);
         Registration registration = agentInstance.assignCertification();
+
         if (agentInstance.isRegistered()) {
-            agentConfigService.saveOrUpdateAgent(agentInstance, username);
+            //TODO : Re-look at below flow to see if we need to set the cookie and save back agent into db
+            Agent agent = agentInstance.getAgent();
+            generateCookieForAgentIfAgentDoesNotHaveACookie(agent);
+
+            saveOrUpdateAgentInstance(agentInstance);
+            bombIfAgentHasErrors(agent);
             LOGGER.debug("New Agent approved {}", agentRuntimeInfo);
         }
+
         return registration;
     }
 
+    private void generateCookieForAgentIfAgentDoesNotHaveACookie(Agent agent) {
+        if (agent.getCookie() == null) {
+            String cookie = uuidGenerator.randomUuid();
+            agent.setCookie(cookie);
+        }
+    }
+
+    private void bombIfAgentHasErrors(Agent agent) {
+        if (agent.hasErrors()) {
+            List<ConfigErrors> errors = ErrorCollector.getAllErrors(agent);
+
+            throw new GoConfigInvalidException(null, new AllConfigErrors(errors));
+        }
+    }
+
+    public void updateAgentApprovalStatus(final String uuid, final Boolean isDenied) {
+        AgentInstance agentInstance = agentInstances.findAgent(uuid);
+
+        Agent agent = (agentInstance.isRegistered() ? agentInstance.getAgent() : null);
+        bombIfNull(agent, "Unable to update agent approval status; Agent [" + uuid + "] not found.");
+
+        agent.setDisabled(isDenied);
+        saveOrUpdate(agent);
+    }
+
+    private void saveOrUpdateAgentInstance(AgentInstance agentInstance) {
+        Agent agent = agentInstance.getAgent();
+        if (!agentInstance.isRegistered()) {
```

Comment:
Yes true.

Removed AgentService.saveOrUpdateAgentInstance() and replaced call to that with AgentService.saveOrUpdate() call,
    
Because all the method (AgentService.saveOrUpdateAgentInstance()) did was,    

- In AgentService.requestRegistration(), if an agent is registered, then save it in db by calling AgentService.saveOrUpdate()
- Otherwise save the approval status of an agent
- The method AgentService.saveOrUpdateAgentInstance() was called from only one place and called only when an agent is registered
- Which means, the otherwise (else part) wherein agent approval status is saved was a stale code,

    - So removed that and then what was left in the method AgentService.saveOrUpdateAgentInstance() was a call to AgentService.saveOrUpdate()
    - Hence replaced AgentService.saveOrUpdateAgentInstance() with call to AgentService.saveOrUpdate() from AgentService.requestRegistration()